presto-docs/src/main/sphinx/optimizer/history-based-optimization.rst

```diff
@@ -0,0 +1,97 @@
+                         - RemoteSource[2] => [orderpriority:varchar(15), count_8:bigint, $hashvalue_9:bigint]
+
+    
+Optimizations using HBO
+-----------------------
+
+DetermineJoinDistributionType & DetermineSemiJoinDistributionType
+-----------------------------------------------------------------
+These two optimizations decide whether to do broadcast or repartition for a join. They use the size of the probe and build input in optimization.
+* Data size recorded from history queries will be used when HBO is enabled.
+* Statistics from cost models are used when HBO statistics are not available or HBO is disabled.
+
+ReorderJoins
+------------
+This optimization reorders the join order based on the size of input and output. Data size recorded from history queries will be used when HBO is enabled.
+
+PushPartialAggregationThroughExchange
+-------------------------------------
+This optimization decides whether to split an aggregation into partial and final aggregations. HBO has special session properties to control it.
+* Set `track_partial_aggregation_history` to true to track the output size of the partial aggregation node.
+* Set `use_partial_aggregation_history` to true to use the partial aggregation node statistics to decide whether to split aggregation. The track of partial aggregation statistics
+targets the pattern we found in production queries, where the final aggregation node is cardinality reducing but partial aggregation is not.
+When `use_partial_aggregation_history` is not enabled or partial aggregation statistics are not available, it will fall back to using the final aggregation statistics.
+Note: When the optimizer disables partial aggregation, there are no statistics about partial aggregation and the partial aggregation statistics are unavailable.
+
+ScaledWriterRule
+----------------
+Scaled writer supports dynamically increasing the number of file write tasks, so as to avoid writing out too many small files. By default it starts with one write task.
+In HBO, the number of tasks used for writing files is recorded as history. ScaledWriterRule decides the number of tasks to start with based on this information.
+It will start with half of the number of write tasks recorded in HBO , because scaled writer only increases the number of write tasks and it will never decrease if we use
+exactly the same number of tasks from history runs. This optimization can be enabled by session property `enable_history_based_scaled_writer`.
+
+RandomizeNullKeyInOuterJoin
+---------------------------
+RandomizeNullKeyInOuterJoin is used to mitigate skew of NULL values in outer joins by rewriting NULL keys to non null keys which will never match.
+It benefits queries with outer join where the join key has skew on NULL values.
+In HBO, the number of NULL keys and total join keys are tracked for join nodes; this optimization will be enabled when the portion of NULL keys exceeds the following thresholds:
+.
```

Comment:
```suggestion

```
Need a line after this text for the two items in the unordered list to build correctly.